### ISSUE TITLE
Field promotion, this vs super

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#50948]
```dart
  test_class_field_read_propertyAccess_super_checkLocal() async {
    await assertNoErrorsInCode('''
class C {
  final int? _foo;
  C(this._foo);
}

class D extends C {
  final int? _foo;
  D(this._foo) : super(_foo);

  void f() {
    if (_foo != null) {
      _foo; // 1
      super._foo; // 2
    }
    _foo; // 3
    super._foo; // 4
  }
}
''');
    print(_resolvedNodeText(findNode.simple('_foo; // 1')));
    print(_resolvedNodeText(findNode.propertyAccess('_foo; // 2')));
    print(_resolvedNodeText(findNode.simple('_foo; // 3')));
    print(_resolvedNodeText(findNode.propertyAccess('_foo; // 4')));
  }
```

Produces promoted types for both `// 1` and `// 2`.
```
SimpleIdentifier
  token: _foo
  staticElement: self::@class::D::@getter::_foo
  staticType: int

PropertyAccess
  target: SuperExpression
    superKeyword: super
    staticType: D
  operator: .
  propertyName: SimpleIdentifier
    token: _foo
    staticElement: self::@class::C::@getter::_foo
    staticType: int
  staticType: int

SimpleIdentifier
  token: _foo
  staticElement: self::@class::D::@getter::_foo
  staticType: int?

PropertyAccess
  target: SuperExpression
    superKeyword: super
    staticType: D
  operator: .
  propertyName: SimpleIdentifier
    token: _foo
    staticElement: self::@class::C::@getter::_foo
    staticType: int?
  staticType: int?
```

I don't see a project for this, so I assigned `Patterns` as here as well.

Discovered via `co19/LanguageFeatures/Private-fields-promotion/promotion_A01_t05`

Paul, does this look as a bug in the analyzer integration of this feature, or do you think that this is a bug in the flow analysis?

**Comment by stereotype441** [comment]
Unassigning myself because I'm not actively working on field promotion right now; I will likely pick it up again later in the year.

**Comment by lrhn** [comment]
It does look like a bug in any case. The `super._foo` is not checked by the `if (_foo != null)`, so even though it's a non-virtual getter invocation, and therefore should be safe against being overridden, so it *could* be promoted, there is no promoting check for it here.

**Comment by stereotype441** [comment]
Fixed in 70e7adff19d927e3375d1ba11350a40b79aafd54.
